mongo/integration/log_helpers_test.go

```diff
@@ -0,0 +1,119 @@
+// Copyright (C) MongoDB, Inc. 2023-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package integration
+
+import (
+	"context"
+	"fmt"
+
+	"go.mongodb.org/mongo-driver/internal/logger"
+	"go.mongodb.org/mongo-driver/mongo/integration/mtest"
+)
+
+type testLogSink struct {
+	logs       chan func() (int, string, []interface{})
+	bufferSize int
+	logsCount  int
+	errsCh     chan error
+}
+
+type logValidator func(order int, lvl int, msg string, kv ...interface{}) error
+
+func newTestLogSink(ctx context.Context, mt *mtest.T, bufferSize int, validator logValidator) *testLogSink {
+	mt.Helper()
+
+	sink := &testLogSink{
+		logs:       make(chan func() (int, string, []interface{}), bufferSize),
+		errsCh:     make(chan error, bufferSize),
+		bufferSize: bufferSize,
+	}
+
+	go func() {
+		order := 0
+		for log := range sink.logs {
+			select {
+			case <-ctx.Done():
+				sink.errsCh <- ctx.Err()
+
+				return
+			default:
+			}
+
+			level, msg, args := log()
+			if err := validator(order, level, msg, args...); err != nil {
+				sink.errsCh <- fmt.Errorf("invalid log at position %d, level %d, and msg %q: %v", order,
```

Comment:
Use the `%w` format verb instead of `%v` so `fmt.Errorf` creates a wrapped error.